File: frontend/import-sheet.js
/* Chord-sheet import, in the browser.
 *
 * A faithful port of pipeline/import_sheet.py (+ the merge step) so the
 * hosted version can import a pasted tab without the PC. Both must produce
 * the same placements - see cloud/compare_import.py for the parity check.
 */
(function () {
  const NOTES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
  const FLAT_TO_SHARP = { Db: "C#", Eb: "D#", Gb: "F#", Ab: "G#", Bb: "A#",
                          Cb: "B", Fb: "E" };
  const CHORD_RE =
    /^[A-G][#b♯♭]?(?:maj|min|m|M|dim|aug|sus|add)?[0-9]*(?:(?:sus|add|maj|dim|aug|b|#)[0-9]*)*(?:\/[A-G][#b♯♭]?)?$/;
  const IGNORE = new Set(["N.C.", "NC", "|", "-", "–", "%", "x2", "x3", "x4",
                          "(x2)", "(x3)"]);
  const NOTE_SYMBOL = "♪";
  const MERGE_GAP = 3.0;

  const norm = t => t.toLowerCase().replace(/[^a-z0-9']/g, "");

  function isChordToken(tok) {
    const t = tok.replace(/^[(,.|]+|[),.|]+$/g, "");
    if (!t || IGNORE.has(t)) return false;
    return CHORD_RE.test(t) && t.length <= 10;
  }

  function classifyLine(line) {
    if (/^\s*\[[^\]]+\]\s*$/.test(line)) return "section";
    const toks = line.split(/\s+/).filter(t => t && !IGNORE.has(t));
    if (!toks.length) return "blank";
    const chordish = toks.filter(isChordToken).length;
    return chordish >= Math.max(1, Math.round(0.75 * toks.length)) ? "chords" : "lyric";
  }

  function tokensWithCols(line) {
    const out = [];
    const re = /\S+/g;
    let m;
    while ((m = re.exec(line)) !== null) out.push([m.index, m[0]]);
    return out;
  }

  function parseSheet(text) {
    const capoM = /capo[^0-9]{0,10}(\d{1,2})/i.exec(text);
    const capo = capoM ? parseInt(capoM[1], 10) : null;
    const pairs = [];
    let pending = null, unanchored = 0;
    for (const raw of text.split(/\r?\n/)) {
      const line = raw.replace(/\s+$/, "");
      const kind = classifyLine(line);
      if (kind === "chords") {
        if (pending) unanchored++;
        pending = tokensWithCols(line)
          .filter(([, t]) => isChordToken(t))
          .map(([c, t]) => [c, t.replace(/^[(,.|]+|[),.|]+$/g, "")]);
      } else if (kind === "lyric") {
        pairs.push({ chords: pending || [], lyric: line });
        pending = null;
      } else {
        if (pending) unanchored++;
        pending = null;
      }
    }
    if (pending) unanchored++;
    return { pairs, capo, unanchored };
  }

  /* ---- difflib-compatible sequence matching ---- */

  // Longest matching block in a[alo:ahi] vs b[blo:bhi]; ties go to the
  // earliest position, matching Python's difflib.
  function longestMatch(a, b, alo, ahi, blo, bhi) {
    let besti = alo, bestj = blo, bestsize = 0;
    let j2len = new Map();
    for (let i = alo; i < ahi; i++) {
      const newj2len = new Map();
      for (let j = blo; j < bhi; j++) {
        if (a[i] !== b[j]) continue;
        const k = (j2len.get(j - 1) || 0) + 1;
        newj2len.set(j, k);
        if (k > bestsize) { besti = i - k + 1; bestj = j - k + 1; bestsize = k; }
      }
      j2len = newj2len;
    }
    return [besti, bestj, bestsize];
  }

  function matchingBlocks(a, b) {
    const queue = [[0, a.length, 0, b.length]];
    const blocks = [];
    while (queue.length) {
      const [alo, ahi, blo, bhi] = queue.pop();
      const [i, j, k] = longestMatch(a, b, alo, ahi, blo, bhi);
      if (!k) continue;
      blocks.push([i, j, k]);
      if (alo < i && blo < j) queue.push([alo, i, blo, j]);
      if (i + k < ahi && j + k < bhi) queue.push([i + k, ahi, j + k, bhi]);
    }
    blocks.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
    return blocks;
  }

  function ratio(a, b) {
    if (!a.length && !b.length) return 1;
    let matches = 0;
    for (const [, , k] of matchingBlocks(a, b)) matches += k;
    return (2 * matches) / (a.length + b.length);
  }

  /* ---- chord helpers ---- */

  function splitChord(name) {
    const m = /^([A-G][#b]?)(.*)$/.exec(name || "");
    if (!m) return [null, null];
    const root = FLAT_TO_SHARP[m[1]] || m[1];
    const i = NOTES.indexOf(root);
    return i < 0 ? [null, null] : [i, m[2]];
  }

  function transpose(name, semitones) {
    if (!semitones) return name;
    const [idx, suffix] = splitChord(name);
    if (idx === null) return name;
    let suf = suffix;
    if (suf.includes("/")) {
      const [body, bass] = suf.split("/");
      const [bidx, bsuf] = splitChord(bass);
      if (bidx !== null)
        suf = `${body}/${NOTES[(bidx + semitones + 12) % 12]}${bsuf}`;
    }
    return NOTES[(((idx + semitones) % 12) + 12) % 12] + suf;
  }

  const isMinor = s => s.startsWith("m") && !s.startsWith("maj");

  function keyOffset(imported, detected) {
    const scores = new Map();
    for (const a of imported) {
      const [ai, asuf] = splitChord(a.chord);
      if (ai === null) continue;
      for (const b of detected) {
        const lo = Math.max(a.start_time, b.start_time);
        const hi = Math.min(a.end_time, b.end_time);
        if (hi <= lo) continue;
        const [bi, bsuf] = splitChord(b.chord);
        if (bi === null || isMinor(asuf) !== isMinor(bsuf)) continue;
        const d = (((bi - ai) % 12) + 12) % 12;
        scores.set(d, (scores.get(d) || 0) + (hi - lo));
      }
    }
    let total = 0, best = 0, bestOff = 0;
    for (const [k, v] of scores) { total += v; if (v > best) { best = v; bestOff = k; } }
    return total ? [bestOff, best / total] : [0, 0];
  }

  /* ---- matching + assembly ---- */

  const LOOKAHEAD = 80;

  function matchPairs(pairs, words) {
    const wnorm = words.map(w => norm(w.word));
    const events = [], covered = [], unmatched = [], previewLines = [];
    let matchedLines = 0, cursor = 0;
    let bestAttempt = { score: 0, lyric: "", near: "" };

    for (const pair of pairs) {
      const toks = tokensWithCols(pair.lyric);
      const snorm = toks.map(([, t]) => norm(t));
      if (!snorm.length) continue;
      const short = snorm.length < 3;
      const threshold = short ? 0.65 : 0.5;
      const maxStart = words.length - Math.max(2, Math.floor(snorm.length / 2));

      const scan = (lo, hi) => {
        let bestAnyS = 0, bestAnyI = -1, hitS = 0, hitI = -1, firstHit = -1;
        for (let start = lo; start < Math.max(lo + 1, hi); start++) {
          const window = wnorm.slice(start, start + snorm.length + 2);
          const sc = ratio(snorm, window);
          if (sc > bestAnyS) { bestAnyS = sc; bestAnyI = start; }
          if (sc >= threshold) {
            if (firstHit < 0) firstHit = start;
            if (sc > hitS) { hitS = sc; hitI = start; }
          }
          if (firstHit >= 0 && start > firstHit + snorm.length + 2) break;
        }
        return hitI >= 0 ? [hitS, hitI] : [bestAnyS, bestAnyI];
      };

      let [score, start] = scan(cursor, Math.min(maxStart, cursor + LOOKAHEAD));
      if (score < threshold && pair.chords.length && !short) {
        const [g, gi] = scan(Math.min(maxStart, cursor + LOOKAHEAD), maxStart);
        if (g > score) { score = g; start = gi; }
      }
      if (score > bestAttempt.score)
        bestAttempt = { score, lyric: pair.lyric.trim().slice(0, 60),
                        near: wnorm.slice(Math.max(0, start), Math.max(0, start) + 8).join(" ") };
      if (start < 0 || score < threshold) {
        if (pair.chords.length) unmatched.push(pair.lyric.trim().slice(0, 60));
        continue;
      }

      matchedLines++;
      const spanLen = Math.min(snorm.length + 2, words.length - start);
      const window = wnorm.slice(start, start + spanLen);
      const mapping = new Map();
      for (const [ai, bi, k] of matchingBlocks(snorm, window))
        for (let x = 0; x < k; x++) mapping.set(ai + x, start + bi + x);

      const placements = [];
      for (const [col, chordRaw] of pair.chords) {
        let tokIdx = 0;
        toks.forEach(([tcol], i) => { if (tcol <= col) tokIdx = i; });
        let widx = mapping.get(tokIdx);
        if (widx === undefined)
          widx = start + Math.min(spanLen - 1,
                   Math.round(tokIdx * spanLen / Math.max(1, snorm.length)));
        const chord = chordRaw.replace(/♯/g, "#").replace(/♭/g, "b");
        events.push({ t: words[widx].start_time, chord, word: words[widx].word });
        placements.push([widx - start, chord]);
      }
      if (pair.chords.length)
        previewLines.push({ words: words.slice(start, start + spanLen).map(w => w.word),
                            placements });
      covered.push([words[start].start_time - 0.5,
                    words[Math.min(start + spanLen, words.length) - 1].end_time + 0.5]);
      cursor = start + Math.max(1, spanLen - 2);
    }

    return { events, covered, report: {
      chord_lines: pairs.filter(p => p.chords.length).length,
      lyric_lines: pairs.length,
      matched_lines: matchedLines,
      chords_placed: events.length,
      unmatched: unmatched.slice(0, 8),
      best_attempt: bestAttempt,
      preview_lines: previewLines.slice(0, 50),
    } };
  }

  function eventsToSegments(events) {
    const sorted = [...events].sort((a, b) => a.t - b.t);
    const merged = [];
    for (const e of sorted) {
      const last = merged[merged.length - 1];
      if (last && e.chord === last.chord && e.t - last.t < 12) continue;
      merged.push(e);
    }
    return merged.map((e, i) => ({
      chord: e.chord,
      start_time: +e.t.toFixed(2),
      end_time: +((i + 1 < merged.length ? merged[i + 1].t : e.t + 4.0).toFixed(2)),
    }));
  }

  function buildSegments(segments, covered, detected, offset) {
    const inCovered = t => covered.some(([a, b]) => a <= t && t <= b);
    const kept = detected.filter(s => !inCovered(s.start_time))
                         .map(s => ({ ...s, chord: transpose(s.chord, -offset) }));
    return [[...segments, ...kept].sort((a, b) => a.start_time - b.start_time),
            segments.length, kept.length];
  }

  // Port of pipeline/merge.py
  function mergeWordsChords(words, chords, gap = MERGE_GAP) {
    const out = words.map((w, i) => ({ ...w, chord: null, wi: i }));
    const pseudo = [];
    for (const seg of [...chords].sort((a, b) => a.start_time - b.start_time)) {
      const target = out.find(w => w.start_time >= seg.start_time);
      if (target && target.start_time - seg.start_time <= gap) {
        target.chord = seg.chord;
        target.chord_time = seg.start_time;
      } else {
        pseudo.push({ word: NOTE_SYMBOL, start_time: seg.start_time,
                      end_time: seg.end_time, chord: seg.chord,
                      chord_time: seg.start_time, instrumental: true });
      }
    }
    const combined = [...out, ...pseudo].sort((a, b) => a.start_time - b.start_time);
    let group = -1, prevInstrumental = false;
    for (const w of combined) {
      if (w.instrumental) {
        if (!prevInstrumental) group++;
        w.line = `i${group}`;
      }
      prevInstrumental = !!w.instrumental;
    }
    return combined;
  }

  function importSheet(text, words, detectedChords) {
    const { pairs, capo, unanchored } = parseSheet(text);
    if (!pairs.some(p => p.chords.length))
      throw new Error("no chord-over-lyric lines found - paste the plain text " +
                      "of a chord sheet (chords positioned above lyrics)");
    const { events, covered, report } = matchPairs(pairs, words);
    if (!events.length) {
      const ba = report.best_attempt;
      throw new Error("none of the sheet's lyrics appear in this song's " +
        `transcript (best line similarity ${Math.round(ba.score * 100)}%). ` +
        `Closest miss: sheet "${ba.lyric}" vs transcript "${ba.near}". ` +
        "Double-check that this tab is for this exact song and version.");
    }
    const imported = eventsToSegments(events);
    let [offset, confidence] = keyOffset(imported, detectedChords);
    if (confidence < 0.45) offset = 0;
    const [segments, importedCount, keptCount] =
      buildSegments(imported, covered, detectedChords, offset);
    Object.assign(report, {
      key_offset: offset,
      key_confidence: +confidence.toFixed(2),
      sheet_capo: capo,
      unanchored_chord_lines: unanchored,
      imported_segments: importedCount,
      kept_detected: keptCount,
      total_segments: segments.length,
    });
    return { segments, report };
  }

  window.ImportSheet = { importSheet, mergeWordsChords, transpose };
})();

File: pipeline/import_sheet.py
"""Align a pasted chord sheet (chords positioned above lyric lines) to the
transcript's word timestamps.

The sheet's LYRICS are only used for matching - the app keeps its own
transcribed lyrics. Each chord is attached to the timestamp of the
transcript word it sits above, via fuzzy line matching. Sections the sheet
can't anchor (intros, solos) keep the app's detected chords.
"""
import collections
import difflib
import re

# Root + common quality spellings + optional slash bass. Deliberately strict:
# a lyric word must not pass ("Am" passes by design - context filters it).
CHORD_RE = re.compile(
    r"^[A-G][#b♯♭]?"
    r"(?:maj|min|m|M|dim|aug|sus|add)?[0-9]*"
    r"(?:(?:sus|add|maj|dim|aug|b|#)[0-9]*)*"
    r"(?:/[A-G][#b♯♭]?)?$")
IGNORE_TOKENS = {"N.C.", "NC", "|", "-", "–", "%", "x2", "x3", "x4", "(x2)", "(x3)"}
SECTION_RE = re.compile(r"^\s*\[[^\]]+\]\s*$")
CAPO_RE = re.compile(r"capo[^0-9]{0,10}(\d{1,2})", re.IGNORECASE)


NOTES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
FLAT_TO_SHARP = {"Db": "C#", "Eb": "D#", "Gb": "F#", "Ab": "G#", "Bb": "A#",
                 "Cb": "B", "Fb": "E"}


def split_chord(name: str):
    """-> (root index 0-11, suffix) or (None, None) for N/unparseable."""
    m = re.match(r"^([A-G][#b]?)(.*)$", name or "")
    if not m:
        return None, None
    root = FLAT_TO_SHARP.get(m.group(1), m.group(1))
    if root not in NOTES:
        return None, None
    return NOTES.index(root), m.group(2)


def transpose(name: str, semitones: int) -> str:
    """Shift a chord (and any slash bass) by semitones."""
    if not semitones:
        return name
    idx, suffix = split_chord(name)
    if idx is None:
        return name
    if "/" in suffix:
        body, _, bass = suffix.partition("/")
        bidx, bsuf = split_chord(bass)
        if bidx is not None:
            suffix = f"{body}/{NOTES[(bidx + semitones) % 12]}{bsuf}"
    return NOTES[(idx + semitones) % 12] + suffix


def is_minor(suffix: str) -> bool:
    return suffix.startswith("m") and not suffix.startswith("maj")


def key_offset(imported, detected):
    """How many semitones the RECORDING sounds above the SHEET.

    A capo'd recording is detected at sounding pitch while the sheet writes
    the shapes you fret, so the two disagree by a constant. Votes are
    weighted by how long the two segments overlap (a chord held for 4s says
    more than a passing one) and only compare like qualities, which shrugs
    off the timing jitter of chord-above-word placement.
    -> (offset, confidence 0-1)"""
    scores = collections.Counter()
    for a in imported:
        ai, asuf = split_chord(a["chord"])
        if ai is None:
            continue
        for b in detected:
            lo = max(a["start_time"], b["start_time"])
            hi = min(a["end_time"], b["end_time"])
            if hi <= lo:
                continue
            bi, bsuf = split_chord(b["chord"])
            if bi is None or is_minor(asuf) != is_minor(bsuf):
                continue
            scores[(bi - ai) % 12] += hi - lo
    total = sum(scores.values())
    if not total:
        return 0, 0.0
    off, best = scores.most_common(1)[0]
    return off, best / total


def normalise_token(tok: str) -> str:
    return re.sub(r"[^a-z0-9']", "", tok.lower())


def is_chord_token(tok: str) -> bool:
    t = tok.strip("(),.|")
    if not t or t in IGNORE_TOKENS:
        return False
    return bool(CHORD_RE.match(t)) and len(t) <= 10


def classify_line(line: str) -> str:
    if SECTION_RE.match(line):
        return "section"
    toks = [t for t in line.split() if t not in IGNORE_TOKENS]
    if not toks:
        return "blank"
    chordish = sum(1 for t in toks if is_chord_token(t))
    # A chord line is nearly all chord tokens; lyric lines with words like
    # "A" or "Am" don't qualify because their other words fail the regex.
    return "chords" if chordish >= max(1, round(0.75 * len(toks))) else "lyric"


def parse_sheet(text: str):
    """-> (pairs, capo, unanchored_chord_lines)
    pairs: [{chords: [(col, chord)], lyric: str}] in sheet order."""
    capo_m = CAPO_RE.search(text)
    capo = int(capo_m.group(1)) if capo_m else None
    pairs, pending, unanchored = [], None, 0
    for raw in text.splitlines():
        line = raw.rstrip()
        kind = classify_line(line)
        if kind == "chords":
            if pending:
                unanchored += 1
            pending = [(m.start(), m.group().strip("(),.|"))
                       for m in re.finditer(r"\S+", line)
                       if is_chord_token(m.group())]
        elif kind == "lyric":
            pairs.append({"chords": pending or [], "lyric": line})
            pending = None
        elif kind in ("section", "blank"):
            if pending:
                unanchored += 1
            pending = None
    if pending:
        unanchored += 1
    return pairs, capo, unanchored


def match_pairs(pairs, words):
    """Fuzzy-align each sheet lyric line to the transcript (monotonic), and
    place that line's chords on transcript word timestamps.
    -> (events, covered_spans, report)"""
    wnorm = [normalise_token(w["word"]) for w in words]
    events, covered, unmatched = [], [], []
    matched_lines = 0
    cursor = 0
    best_attempt = {"score": 0.0, "lyric": "", "near": ""}
    preview_lines = []

    LOOKAHEAD = 80   # words; verses repeat in songs, so prefer the nearest
    for pair in pairs:
        toks = [(m.start(), m.group()) for m in re.finditer(r"\S+", pair["lyric"])]
        snorm = [normalise_token(t) for _, t in toks]
        if not snorm:
            continue
        short = len(snorm) < 3
        threshold = 0.65 if short else 0.5
        max_start = len(words) - max(2, len(snorm) // 2)

        def scan(lo, hi):
            """Peak of the FIRST region scoring >= threshold (choruses repeat;
            the nearest occurrence is the right one, not the best-scoring one
            later in the song). Falls back to the best miss for diagnostics."""
            best_any_s, best_any_i = 0.0, -1
            hit_s, hit_i, first_hit = 0.0, -1, -1
            for start in range(lo, max(lo + 1, hi)):
                window = wnorm[start:start + len(snorm) + 2]
                score = difflib.SequenceMatcher(None, snorm, window).ratio()
                if score > best_any_s:
                    best_any_s, best_any_i = score, start
                if score >= threshold:
                    if first_hit < 0:
                        first_hit = start
                    if score > hit_s:
                        hit_s, hit_i = score, start
                if first_hit >= 0 and start > first_hit + len(snorm) + 2:
                    break   # past the first acceptable region: stop before repeats
            return (hit_s, hit_i) if hit_i >= 0 else (best_any_s, best_any_i)

        # Local window first: the right occurrence of a repeated line is the
        # next one, not a higher-scoring copy in a later verse/chorus.
        best_score, best_start = scan(cursor, min(max_start, cursor + LOOKAHEAD))
        if best_score < threshold and pair["chords"] and not short:
            # only chord-bearing full lines may jump ahead (e.g. the sheet
            # includes a verse the transcript timeline reaches much later)
            g_score, g_start = scan(min(max_start, cursor + LOOKAHEAD), max_start)
            if g_score > best_score:
                best_score, best_start = g_score, g_start
        if pair["chords"] and best_score > best_attempt["score"]:
            near = " ".join(w["word"] for w in
                            words[best_start:best_start + len(snorm)]) if best_start >= 0 else ""
            best_attempt = {"score": best_score,
                            "lyric": pair["lyric"].strip()[:60], "near": near[:60]}
        if best_start < 0 or best_score < threshold:
            if pair["chords"]:   # chordless chatter lines fail silently
                unmatched.append(pair["lyric"].strip()[:60])
            continue
        matched_lines += 1
        span_len = min(len(snorm) + 2, len(words) - best_start)
        window = wnorm[best_start:best_start + span_len]
        # sheet-token index -> transcript word index, via matching blocks
        mapping = {}
        sm = difflib.SequenceMatcher(None, snorm, window)
        for block in sm.get_matching_blocks():
            for k in range(block.size):
                mapping[block.a + k] = best_start + block.b + k
        line_placements = []
        for col, chord in pair["chords"]:
            tok_idx = 0
            for i, (tcol, _) in enumerate(toks):
                if tcol <= col:
                    tok_idx = i
            widx = mapping.get(tok_idx)
            if widx is None:  # proportional fallback within the span
                widx = best_start + min(span_len - 1,
                                        round(tok_idx * span_len / max(1, len(snorm))))
            chord = chord.replace("♯", "#").replace("♭", "b")
            events.append({"t": words[widx]["start_time"], "chord": chord,
                           "word": words[widx]["word"]})
            line_placements.append([widx - best_start, chord])
        if pair["chords"]:
            preview_lines.append({
                "words": [w["word"] for w in words[best_start:best_start + span_len]],
                "placements": line_placements,   # [word index in line, chord]
            })
        covered.append((words[best_start]["start_time"] - 0.5,
                        words[min(best_start + span_len, len(words)) - 1]["end_time"] + 0.5))
        cursor = best_start + max(1, span_len - 2)

    report = {
        "chord_lines": sum(1 for p in pairs if p["chords"]),
        "lyric_lines": len(pairs),
        "matched_lines": matched_lines,
        "chords_placed": len(events),
        "unmatched": unmatched[:8],
        "best_attempt": best_attempt,
        "preview_lines": preview_lines[:50],
    }
    return events, covered, report


def events_to_segments(events):
    """Collapse chord placements into {chord, start_time, end_time} spans."""
    events = sorted(events, key=lambda e: e["t"])
    merged_events = []
    for e in events:
        if merged_events and e["chord"] == merged_events[-1]["chord"] \
                and e["t"] - merged_events[-1]["t"] < 12:
            continue  # same chord restated shortly after: keep one change
        merged_events.append(e)
    segments = []
    for i, e in enumerate(merged_events):
        end = merged_events[i + 1]["t"] if i + 1 < len(merged_events) else e["t"] + 4.0
        segments.append({"chord": e["chord"],
                         "start_time": round(e["t"], 2),
                         "end_time": round(end, 2)})
    return segments


def build_segments(segments, covered, detected, offset=0):
    """Imported chords own the covered (sung) spans; the app's detected
    chords are kept everywhere else (intro/solo/outro), transposed by
    `offset` so the whole chart stays in the sheet's key."""
    def in_covered(t):
        return any(a <= t <= b for a, b in covered)

    kept = [{**s, "chord": transpose(s["chord"], -offset)}
            for s in detected if not in_covered(s["start_time"])]
    combined = sorted(segments + kept, key=lambda s: s["start_time"])
    return combined, len(segments), len(kept)


def import_sheet(text: str, words: list, detected_chords: list):
    """-> (segments, report). Raises ValueError when nothing usable."""
    pairs, capo, unanchored = parse_sheet(text)
    if not any(p["chords"] for p in pairs):
        raise ValueError("no chord-over-lyric lines found - paste the plain "
                         "text of a chord sheet (chords positioned above lyrics)")
    events, covered, report = match_pairs(pairs, words)
    if not events:
        ba = report["best_attempt"]
        raise ValueError(
            "none of the sheet's lyrics appear in this song's transcript "
            f"(best line similarity {ba['score']:.0%}). Closest miss: sheet "
            f"\"{ba['lyric']}\" vs transcript \"{ba['near']}\". Double-check "
            "that this tab is for this exact song and version.")
    # A capo'd recording is DETECTED at sounding pitch but WRITTEN as shapes
    # in the sheet. Without correcting for it the intro/solo sections (which
    # keep detected chords) would sit a few semitones off the verses.
    imported = events_to_segments(events)
    offset, confidence = key_offset(imported, detected_chords)
    if confidence < 0.45:
        offset = 0
    segments, imported_count, kept_count = build_segments(imported, covered,
                                                          detected_chords, offset)
    report["key_offset"] = offset
    report["key_confidence"] = round(confidence, 2)
    report["sheet_capo"] = capo
    report["unanchored_chord_lines"] = unanchored
    report["imported_segments"] = imported_count
    report["kept_detected"] = kept_count
    report["total_segments"] = len(segments)
    report["sample"] = [{"chord": e["chord"], "word": e["word"],
                         "time": round(e["t"], 1)} for e in events[:10]]
    return segments, report
